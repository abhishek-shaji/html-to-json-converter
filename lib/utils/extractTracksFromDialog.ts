import { TIME_BETWEEN_CHAR_MS } from '../constants/duration';
import { TRACK_TYPE_DIALOGUE } from '../constants/trackTypes';

export const extractTracksFromDialog = (
  dialog: Element,
  initialState: any,
  timestampStartAt: number
) => {
  const trackMap = { ...initialState };
  const children = [...dialog.children];

  for (const childElement of children) {
    const actionType = childElement.getAttribute('data-action-type');
    const target = childElement.getAttribute('data-action-target');
    const actionId = childElement.getAttribute('data-action-id');
    const timestamp =
      timestampStartAt + dialog.innerHTML.indexOf(childElement.outerHTML) * TIME_BETWEEN_CHAR_MS;

    if (!trackMap[actionType]) {
      trackMap[actionType] = [];
    }

    trackMap[actionType].push({ actionId, actionType, timestamp, target });

    childElement.remove();
  }

  const dialogueDuration = dialog.textContent.length * TIME_BETWEEN_CHAR_MS;

  if (!trackMap[TRACK_TYPE_DIALOGUE]) {
    trackMap[TRACK_TYPE_DIALOGUE] = [];
  }

  trackMap[TRACK_TYPE_DIALOGUE].push({
    dialog: dialog.textContent,
    duration: dialogueDuration,
    timestamp: timestampStartAt,
  });

  return { trackMap, dialogueDuration };
};
