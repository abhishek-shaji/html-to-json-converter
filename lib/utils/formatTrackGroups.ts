import { OutputType, TrackType } from '../types/output';

const formatTracks = (trackMap): TrackType[] =>
  Object.keys(trackMap).reduce((accum: TrackType[], type): TrackType[] => {
    return [
      ...accum,
      {
        type,
        actions: trackMap[type],
      },
    ];
  }, []);

export const formatTrackGroups = (trackGroupData): OutputType =>
  Object.keys(trackGroupData).reduce(
    (accum: OutputType, target): OutputType => {
      return {
        trackGroups: [
          ...accum.trackGroups,
          {
            target,
            tracks: formatTracks(trackGroupData[target]),
          },
        ],
      };
    },
    {
      trackGroups: [],
    }
  );
