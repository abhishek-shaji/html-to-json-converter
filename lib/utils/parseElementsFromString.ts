import { JSDOM } from 'jsdom';
import { EmptyFileException } from '../exceptions/EmptyFileException';

export const parseElementsFromString = (html: string): NodeListOf<Element> => {
  if (!html.trim()) {
    throw new EmptyFileException();
  }

  const dom = new JSDOM(html);
  const { document } = dom.window;

  return document.querySelectorAll('div.character, div.dialog');
};
