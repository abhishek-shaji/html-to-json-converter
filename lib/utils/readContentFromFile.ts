import { readFileSync } from 'fs';

export const readContentFromFile = (path: string): string => {
  try {
    return readFileSync(path, 'utf8');
  } catch (err) {
    console.error(err);
  }
};
