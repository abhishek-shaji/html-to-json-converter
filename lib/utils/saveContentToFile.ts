import { writeFileSync, existsSync, mkdirSync } from 'fs';
import * as path from 'path';

const removeFilePart = (dirname) => path.dirname(dirname);

export const saveContentToFile = (file: string, data: string) => {
  const dirname = removeFilePart(file);

  if (!existsSync(dirname)) {
    mkdirSync(dirname, { recursive: true });
  }

  writeFileSync(file, data);
};
