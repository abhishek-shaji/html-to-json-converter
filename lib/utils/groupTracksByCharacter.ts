import { CLASS_NAME_DIALOG } from '../constants/classNames';
import { extractTracksFromDialog } from './extractTracksFromDialog';

export const groupTracksByCharacter = (elements: NodeListOf<Element>) => {
  const targets = {};
  let offset = 0;

  for (const [elementIndex, element] of elements.entries()) {
    const isDialog = element.classList.contains(CLASS_NAME_DIALOG);

    if (isDialog) {
      const character = elements[elementIndex - 1].innerHTML;

      if (!targets[character]) {
        targets[character] = {};
      }

      const { trackMap, dialogueDuration } = extractTracksFromDialog(
        element,
        targets[character],
        offset
      );

      offset = offset + dialogueDuration;

      targets[character] = trackMap;
    }
  }

  return targets;
};
