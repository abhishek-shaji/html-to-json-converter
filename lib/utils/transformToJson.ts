import { groupTracksByCharacter } from './groupTracksByCharacter';
import { formatTrackGroups } from './formatTrackGroups';

export const transformToJson = (elements: NodeListOf<Element>) => {
  const targets = groupTracksByCharacter(elements);
  const formattedData = formatTrackGroups(targets);

  return JSON.stringify(formattedData, null, 2);
};
