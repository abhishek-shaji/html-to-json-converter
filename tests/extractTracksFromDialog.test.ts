import { JSDOM } from 'jsdom';
import { extractTracksFromDialog } from '../lib/utils/extractTracksFromDialog';
import {
  TRACK_TYPE_ANIMATION,
  TRACK_TYPE_DIALOGUE,
  TRACK_TYPE_EMOTION,
} from '../lib/constants/trackTypes';

const mockHtmlWithoutActions = `
<div class="dialog">Test Dialog</div>
`;

const mockHtmlWithActions = `
<div class="dialog">Test Dialog with actions <span class="text__action" data-action-type="EM" data-action-target="CH_genius" data-action-id="EM_Laughing">Laughing</span>test<span class="text__action" data-action-type="EM" data-action-target="CH_lucy" data-action-id="EM_Laughing">Laughing</span>test<span class="text__action" data-action-type="AN" data-action-target="CH_genius" data-action-id="AN_Wave">Wave</span>Test dialog?</div>
`;

describe('extractTracksFromDialog', () => {
  it('should properly extract dialog data fom dialog element that contains no actions', () => {
    const { window } = new JSDOM(mockHtmlWithoutActions);

    const dialogElement = window.document.querySelector('.dialog');

    const response = extractTracksFromDialog(dialogElement, {}, 0);

    expect(response).toMatchSnapshot();
    expect(response.dialogueDuration).toEqual(770);
    expect(response.trackMap[TRACK_TYPE_DIALOGUE][0].dialog).toEqual('Test Dialog');
    expect(response.trackMap[TRACK_TYPE_DIALOGUE][0].timestamp).toEqual(0);
  });

  it('should extract dialog and action data fom dialog element that contains multiple actions', () => {
    const { window } = new JSDOM(mockHtmlWithActions);

    const dialogElement = window.document.querySelector('.dialog');

    const response = extractTracksFromDialog(dialogElement, {}, 0);

    expect(response).toMatchSnapshot();
    expect(response.dialogueDuration).toEqual(3150);
    expect(response.trackMap[TRACK_TYPE_DIALOGUE].length).toEqual(1);
    expect(response.trackMap[TRACK_TYPE_EMOTION].length).toEqual(2);
    expect(response.trackMap[TRACK_TYPE_ANIMATION].length).toEqual(1);
  });

  it('should properly calculate timestamp when a dialog element with emotion and animation is passed', () => {
    const { window } = new JSDOM(mockHtmlWithActions);

    const dialogElement = window.document.querySelector('.dialog');

    const response = extractTracksFromDialog(dialogElement, {}, 0);

    expect(response).toMatchSnapshot();
    expect(response.dialogueDuration).toEqual(3150);
    expect(response.trackMap[TRACK_TYPE_DIALOGUE][0].timestamp).toEqual(0);
    expect(response.trackMap[TRACK_TYPE_EMOTION][0].timestamp).toEqual(1750);
    expect(response.trackMap[TRACK_TYPE_EMOTION][1].timestamp).toEqual(2030);
    expect(response.trackMap[TRACK_TYPE_ANIMATION][0].timestamp).toEqual(2310);
  });

  it('should properly calculate timestamp when a dialog element with emotion and animation is passed with offset', () => {
    const { window } = new JSDOM(mockHtmlWithActions);

    const dialogElement = window.document.querySelector('.dialog');

    const response = extractTracksFromDialog(dialogElement, {}, 1000);

    expect(response).toMatchSnapshot();
    expect(response.dialogueDuration).toEqual(3150);
    expect(response.trackMap[TRACK_TYPE_DIALOGUE][0].timestamp).toEqual(1000);
    expect(response.trackMap[TRACK_TYPE_EMOTION][0].timestamp).toEqual(2750);
    expect(response.trackMap[TRACK_TYPE_EMOTION][1].timestamp).toEqual(3030);
    expect(response.trackMap[TRACK_TYPE_ANIMATION][0].timestamp).toEqual(3310);
  });
});
