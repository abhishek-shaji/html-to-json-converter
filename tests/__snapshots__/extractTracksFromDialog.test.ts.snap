// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTracksFromDialog It should extract dialog and action data fom dialog element that contains multiple actions 1`] = `
{
  "dialogueDuration": 3150,
  "trackMap": {
    "AN": [
      {
        "actionId": "AN_Wave",
        "actionType": "AN",
        "target": "CH_genius",
        "timestamp": 2310,
      },
    ],
    "DI": [
      {
        "dialog": "Test Dialog with actions testtestTest dialog?",
        "duration": 3150,
        "timestamp": 0,
      },
    ],
    "EM": [
      {
        "actionId": "EM_Laughing",
        "actionType": "EM",
        "target": "CH_genius",
        "timestamp": 1750,
      },
      {
        "actionId": "EM_Laughing",
        "actionType": "EM",
        "target": "CH_lucy",
        "timestamp": 2030,
      },
    ],
  },
}
`;

exports[`extractTracksFromDialog It should properly calculate timestamp when a dialog element with emotion and animation is passed 1`] = `
{
  "dialogueDuration": 3150,
  "trackMap": {
    "AN": [
      {
        "actionId": "AN_Wave",
        "actionType": "AN",
        "target": "CH_genius",
        "timestamp": 2310,
      },
    ],
    "DI": [
      {
        "dialog": "Test Dialog with actions testtestTest dialog?",
        "duration": 3150,
        "timestamp": 0,
      },
    ],
    "EM": [
      {
        "actionId": "EM_Laughing",
        "actionType": "EM",
        "target": "CH_genius",
        "timestamp": 1750,
      },
      {
        "actionId": "EM_Laughing",
        "actionType": "EM",
        "target": "CH_lucy",
        "timestamp": 2030,
      },
    ],
  },
}
`;

exports[`extractTracksFromDialog It should properly calculate timestamp when a dialog element with emotion and animation is passed with offset 1`] = `
{
  "dialogueDuration": 3150,
  "trackMap": {
    "AN": [
      {
        "actionId": "AN_Wave",
        "actionType": "AN",
        "target": "CH_genius",
        "timestamp": 3310,
      },
    ],
    "DI": [
      {
        "dialog": "Test Dialog with actions testtestTest dialog?",
        "duration": 3150,
        "timestamp": 1000,
      },
    ],
    "EM": [
      {
        "actionId": "EM_Laughing",
        "actionType": "EM",
        "target": "CH_genius",
        "timestamp": 2750,
      },
      {
        "actionId": "EM_Laughing",
        "actionType": "EM",
        "target": "CH_lucy",
        "timestamp": 3030,
      },
    ],
  },
}
`;

exports[`extractTracksFromDialog It should properly extract dialog data fom dialog element that contains no actions 1`] = `
{
  "dialogueDuration": 770,
  "trackMap": {
    "DI": [
      {
        "dialog": "Test Dialog",
        "duration": 770,
        "timestamp": 0,
      },
    ],
  },
}
`;
